src/include/migraphx/op/slice.hpp

```diff
@@ -98,37 +100,75 @@ struct slice
 
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
-        auto input_shape        = inputs[0];
-        auto t                  = input_shape.type();
-        const auto& old_lens    = input_shape.lens();
-        const auto& old_strides = input_shape.strides();
+        check_shapes{inputs, *this, true}.has(1);
+        auto input_shape = inputs[0];
+        auto t           = input_shape.type();
+        // For a static shape, old_lens will be adjusted to a new size
+        // for those axes that are sliced.
+        // For dynamic shape, the adjusted old_lens become the new max values,
+        // while retaining the old mins and opts if possible.
+        std::vector<std::size_t> new_mins;
+        std::vector<std::size_t> new_opts;
+        std::vector<std::size_t> old_lens;
+        std::vector<std::size_t> old_strides;
+        if(input_shape.dynamic())
+        {
+            old_lens = input_shape.max_lens();
+            new_mins = input_shape.min_lens();
+            new_opts = input_shape.opt_lens();
+        }
+        else
+        {
+            old_lens = input_shape.lens();
+            // For static shape (including eval after a dynamic input) the strides are indexed into
+            // the pre-slice array, so they are larger than the apparent size of the resulting
+            // shape.
+            old_strides = input_shape.strides();
+        }
 
         if(std::any_of(
-               axes.begin(), axes.end(), [&](auto i) { return (i >= old_lens.size() and i < 0); }))
+               axes.begin(), axes.end(), [&](auto i) { return (i >= old_lens.size() or i < 0); }))
         {
             MIGRAPHX_THROW("SLICE: input axis " + to_string_range(axes) + " out of range");
         }
 
         if(starts.size() != axes.size() or axes.size() != ends.size())
         {
-            MIGRAPHX_THROW("SLICE: inconsistent sizes");
+            MIGRAPHX_THROW("SLICE: inputs have inconsistent sizes");
         }
 
         std::vector<std::size_t> new_lens = old_lens;
         for(std::size_t i = 0; i < axes.size(); i++)
         {
             auto axis = axes[i];
-            new_lens[axis] =
+            auto sliced_length =
                 fix_index(old_lens, axis, ends[i]) - fix_index(old_lens, axis, starts[i]);
+            // A Numpy indexing convention: a slice size larger than the actual dimension
+            // is legal and is clipped to the axis size
+            new_lens[axis] = std::min(new_lens[axis], sliced_length);
+            if(input_shape.dynamic())
+            {
+                new_mins[axis] = std::min(new_lens[axis], new_mins[axis]);
+                if(new_opts[axis] < new_mins[axis] or new_opts[axis] > new_lens[axis])
+                    new_opts[axis] = 0;
```

Comment:
What would you suggest?  This says that if the opt is still within the min/max range, leave it alone; if not, set it to 0.  It's scarcely even a calcualtion.

Reply:
Here's what I'm thinking:
data shape: [{1, 4}, {4, 10, 8}]
axes: [1]
starts:[2]
ends:[7]

output: [{1, 4}, {2, 5, 5}]

Reply:
In other words, when the new max is less than the opt; you set the opt to the new max value.  To extend the logic, whenever the old opt is outside the new min/max range you clamp it?  That makes sense too, but I don't know if it's what a user would want because I don't know what opts are used for.

Reply:
`opt` is essentially just a suggested value in the range. It should be calculated like a single static dimension; which is often what happens for min and max too.